### PR TITLE
Convert the float health percentage to an int percentage

### LINKED ISSUE #7426: Transforming with low hp causes invulnerability
See https://github.com/OpenRA/OpenRA/pull/7130#issuecomment-72467231.

Steps:
1. Set ```mcv``` [Health](https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/rules/vehicles.yaml#L298) to ```16```.
2. Set ```M1Carbine``` [chance](https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/weapons.yaml#L331) against ```light``` to ```100```.
3. Go ingame and build one mcv and one rifle man.
4. Shoot a single time on the mcv.
5. Deploy and undeploy it.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8254898-transforming-with-low-hp-causes-invulnerability?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by abcdefg30**
As wished by @pchote.

Should fix #7426 as well.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Updated. Thanks @reaperrr.

**Comment by Mailaender** [comment]
I guess we need an upgrade rule for those `map.yaml` changes.

**Comment by pchote** [comment]
I agree.  That is going to require some adjustments to the Map and upgrader code first.

File: OpenRA.Game/Traits/Health.cs [hunk 1]
```diff
@@ -42,8 +43,13 @@ public Health(ActorInitializer init, HealthInfo info)
...
 			MaxHP = info.HP;
 
-			hp = init.Contains<HealthInit>() ? (int)(init.Get<HealthInit, float>() * MaxHP) : MaxHP;
+			hp = init.Contains<HealthInit>() ? init.Get<HealthInit, int>() * MaxHP / 100 : MaxHP;
+			if (hp <= 0)
+				hp = Math.Max(MaxHP / 100, 1);
+
 			DisplayHp = hp;
+
+			Game.Debug("Health: {0}", hp);
```
Comment (pchote) [inline OpenRA.Game/Traits/Health.cs]: This will need to be removed before merging.

File: OpenRA.Game/Traits/Health.cs [hunk 2]
```diff
@@ -169,12 +175,12 @@ public void Tick(Actor self)
 		}
 	}
 
-	public class HealthInit : IActorInit<float>
+	public class HealthInit : IActorInit<int>
 	{
-		[FieldFromYamlKey] readonly float value = 1f;
+		[FieldFromYamlKey] readonly int value = 1;
```
Comment (pchote) [inline OpenRA.Game/Traits/Health.cs]: This should probably default to 100?

File: OpenRA.Mods.Common/Traits/World/BridgeLayer.cs [hunk 3]
```diff
@@ -46,7 +46,7 @@ public void WorldLoaded(World w, WorldRenderer wr)
 			{
 				var bi = w.Map.Rules.Actors[bridge].Traits.Get<BridgeInfo>();
 				foreach (var template in bi.Templates)
-					bridgeTypes.Add(template.First, Pair.New(bridge, template.Second));
+					bridgeTypes.Add(template.First, Pair.New(bridge, (int)(100 * template.Second)));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/BridgeLayer.cs]: It would be nicer to upgrade the bridge rules to do this translation in the yaml instead.

File: OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs [hunk 4]
```diff
@@ -393,7 +393,7 @@ void LoadActors(IniFile file, string section)
 						players.Add(parts[0]);
 
 					var loc = Exts.ParseIntegerInvariant(parts[3]);
-					var health = float.Parse(parts[2], NumberFormatInfo.InvariantInfo) / 256;
+					var health = int.Parse(parts[2], NumberFormatInfo.InvariantInfo) * 100 / 256;
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs]: I don't remember how the legacy maps define health, but if it is as a float then you will need to keep this parsing as a float, and then cast to an int when you're done.

File: OpenRA.Game/Traits/Health.cs (hunk 1 above)
Comment (abcdefg30) [inline OpenRA.Game/Traits/Health.cs]: See the commit description.

File: OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs (hunk 4 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs]: They are stored as integer with 256 being full health. See http://nyerguds.arsaneus-design.com/cncstuff/mappics/campaign-gdi/maps/ for examples.

Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs]: In that case, `int.Parse` is fine.

**Comment by abcdefg30** [comment]
Updated.
I think I'll need some help with the upgrade rule.

It would be good if someone could verify that this fixes #7426.
I could remove the testcase and debug stuff then.

File: OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs (hunk 4 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/LegacyMapImporter.cs]: You can replace it with `Exts.ParseIntegerInvariant` though for consistency.

**Comment by obrakmann** [comment]
Something is wrong with E1s in RA, they almost take out buildings with a single shot.

File: mods/ra/weapons.yaml [hunk 5]
```diff
@@ -324,11 +324,11 @@ M1Carbine:
 		Speed: 1c682
 	Warhead@1Dam: SpreadDamage
 		Spread: 128
-		Damage: 15
+		Damage: 599
```
Comment (WolfGaming) [inline mods/ra/weapons.yaml]: I am pretty sure this isnt supposed to be here.

File: mods/ra/weapons.yaml [hunk 6]
```diff
@@ -324,11 +324,11 @@ M1Carbine:
...
 		DeathType: 2
 		Versus:
 			Wood: 25
-			Light: 30
+			Light: 100
```
Comment (WolfGaming) [inline mods/ra/weapons.yaml]: Also this.

**Comment by abcdefg30** [comment]
> Something is wrong with E1s in RA, they almost take out buildings with a single shot.

> I am pretty sure this isnt supposed to be here.

> Also this.

... See 5659927a4e3c9e7fb8db56a71a5a92c6a9aa0cbe
> It would be good if someone could verify that this fixes #7426.
 I could remove the testcase and debug stuff then.

**Comment by obrakmann** [comment]
This does fix #7426

**Comment by obrakmann** [comment]
No good idea on the upgrade rule, though.  By the time the upgrade rules run, the `ActorInits` have already been parsed, so if there's an invalid value, the utility will have crashed already.  So I suspect an upgrade will somehow have to be done when the field loader runs, probably in the `ActorReference`'s `LoadInit` method.

**Comment by abcdefg30** [comment]
> This does fix #7426 

Thanks, removed the "testcommit".

> So I suspect an upgrade will somehow have to be done when the field loader runs, probably in the  ActorReference 's  LoadInit  method. 

Thanks again. :P I'll have a look at it.

**Comment by Mailaender** [comment]
I guess this needs a `MapFormat` bump and a float to int conversion in the `public Map(string path)` constructor.

**Comment by pchote** [comment]
I have a branch that changes the way that maps are loaded, and provides access to the raw actor MiniYaml so that it can be upgraded.  I'll reference this when I create a pr for those changes after the next release.

**Comment by Unit158** [comment]
What dependencies does this have?

**Comment by abcdefg30** [comment]
> I have a branch that changes the way that maps are loaded, and provides access to the raw actor MiniYaml so that it can be upgraded. I'll reference this pr when I create a pr for those changes after the next release.

**Comment by pchote** [comment]
Note that the patch I'm referring to itself depends on a bunch of currently open prs.  It's otherwise ready to go, so it's just a case of waiting (and reviewing) now.

**Comment by abcdefg30** [comment]
Updated.

**Comment by Phrohdoh** [comment]
```
OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:L41: [SA1009] Invalid spacing around the closing parenthesis.
OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:L43: [SA1009] Invalid spacing around the closing parenthesis.
```

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Activities/Transform.cs [hunk 7]
```diff
@@ -70,10 +70,7 @@ public override Activity Tick(Actor self)
 				var health = self.TraitOrDefault<Health>();
 				if (health != null)
 				{
-					var newHP = (ForceHealthPercentage > 0)
-						? ForceHealthPercentage / 100f
-						: (float)health.HP / health.MaxHP;
-
+					var newHP = (ForceHealthPercentage > 0) ? ForceHealthPercentage : (health.HP * 100) / health.MaxHP;
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Transform.cs]: The parenthesis around `ForceHealthPercentage > 0` can be removed.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 8]
```diff
@@ -1369,7 +1382,17 @@ internal static void UpgradePlayers(int engineVersion, ref List<MiniYamlNode> no
 		internal static void UpgradeActors(int engineVersion, ref List<MiniYamlNode> nodes, MiniYamlNode parent, int depth)
 		{
 			foreach (var node in nodes)
+			{
+				if (engineVersion < 20150430)
+				{
+					if (node.Key == "Health")
+					{
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: These braces are unnecessary and can be removed.

File: mods/cnc/maps/gdi03/map.yaml [hunk 9]
```diff
@@ -834,6 +834,7 @@ Rules:
 		-AirstrikePower:
 		RenderBuilding:
 			Image: hq
+		WithMakeAnimation:
```
Comment (pchote) [inline mods/cnc/maps/gdi03/map.yaml]: Are these `WithMakeAnimation`s supposed to be here?

File: mods/cnc/maps/nod01/map.yaml [hunk 10]
```diff
@@ -296,8 +296,10 @@ Rules:
 		Invulnerable:
 	^CivBuilding:
 		MustBeDestroyed:
+			RequiredForShortGame: true
```
Comment (pchote) [inline mods/cnc/maps/nod01/map.yaml]: These don't seem to be related to the upgrade rule either.  It would be best if these could live in their own commit.

**Comment by pchote** [comment]
Looks good, except for the noted minor issues.  I haven't tested anything ingame.

File: OpenRA.Mods.Common/Activities/Transform.cs (hunk 7 above)
Comment (penev92) [inline OpenRA.Mods.Common/Activities/Transform.cs]: I like them. Some times they make things easier to read.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 8 above)
Comment (penev92) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Same for the outer if's braces.

Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: This is one of the rules explicitly covered in our coding standard.

File: OpenRA.Mods.Common/Activities/Transform.cs (hunk 7 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Transform.cs]: This is one of the rules explicitly covered in our coding standard.

**Comment by abcdefg30** [comment]
Updated.

**Comment by obrakmann** [comment]
Code looks good to me, works in-game as expected, map import is also fine. :+1:

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/b9df1695ce63aeee7c44c4cbcb9470e8b5ab329d%5E...b9df1695ce63aeee7c44c4cbcb9470e8b5ab329d
